### ISSUE TITLE
Make it a library

### ISSUE THREAD

**Issue by furesoft** [10usb/assembler#2]
Could you covert your project to a library to pre evaluate some logic and bundle it as custom Assembler instead of deploying a Definition file?

**Comment by 10usb** [comment]
Do you mean the definition file in the example. So it's easier for more programs to make use of the same? Or do you mean splitting up the cli app and de logic so the assembler logic could be included in other projects?

**Comment by furesoft** [comment]
I mean splitting the project to include the logic in other projects

**Comment by 10usb** [comment]
I see

Usually I start with building libraries and then build an application using those libraries. But for this assembler I wanted 
 it to be a standalone executable. As that makes it easier to just drop it into a project and make use of it. without the hassle of installing it somewhere. It's also done to be a limiting factor by making it less temping to add complex features and keep it simple.

Splitting the project would only remove the Program class to a separate console application. While it's not as conventional in the .NET world to use an executable as a library. It is possible to include the project as if it were a library. The binary format is the same of both.

**Comment by furesoft** [comment]
Hmm, but would be really helpfull for my purpose to evaluate not only a file. Could you add the option to evaluate from a string?

**Comment by 10usb** [comment]
I see you're problem. and think I have  solution for it.

The're no problems introduced with changing the constructor of the Document class to accept a stream instead of a FileInfo.
this would allow you to redirect the output to an in memory variant.

For evaluation of a string there needs to be some form of source reference, otherwise debug info would be lost. This can be tackled with an intermediate class that can supply a reference name and a TextReader as that is what the parser needs.

At the moment only the GlobalInterpreter supports importing and including of other files. And I thinks this can be solved by a dedicated file resolver interface/class. So in default case it will resolve to a files as it is done now. Returning a source reference  that can be used for the next parser. But in your case when you want yo evaluate a string you'd probably also want the import/include instructions to import a string instead of a file. And by this you would be able to add your own adapter as substitute or as middleware.

I hope to have the time tomorrow to make changes.

**Comment by furesoft** [comment]
Yeah, would be awesome. I want to use this for my compiler to allow different platforms easily and use the nice macro system

**Comment by 10usb** [comment]
I've made a concept in the branche string-evaluation. It should be usable for your needs.
The import/include resolver is still a work in progress. Don't like the idea to have yet an other argument being passed from interpreter stage to stage.

This code should be  more or less the start-up to get it working.
```C#
using (Document document = new Document(output)) {
    document.AddImport("Some path for additional imports");

    Router router = new Router();
    router.PushState(new ImportInterpreter(router, document, Trace.Empty));
    using (Parser parser = new Parser(new FileSource("Path to the platform specific language file"), router, Trace.Empty)) {
        parser.Parse();
    }
    router.PopState();

    router.PushState(new GlobalInterpreter(router, document));
    using (Parser parser = new Parser(new StringSource("name of compiled file", "generated code from the compiler"), router, Trace.Empty)) {
        parser.Parse();
    }

    document.Resolve();
}
```

**Comment by furesoft** [comment]
It would be really awesome if you would publish your project on nuget. Then anyone can easily integrate it to make assembler applications and receive new features
